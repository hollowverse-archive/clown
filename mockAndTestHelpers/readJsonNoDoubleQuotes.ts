import { fs } from '@forabi/memfs';

/* Jest, when doing snapshots, wraps strings in double quotes. Since JSON keys and string values
are also wrapped in double quotes, Jest finds itself needing to escape all the double quotes in a
JSON, so the JSON becomes cluttered with \\" \\" \\" \\", making it unreadable. The function
below substitutes double quotes for single quotes when reading JSON content from the mock disk,
so there wouldn't be double-quote escaping issues. */
export function readJsonNoDoubleQuotes(filePathInDisk: string) {
  // tslint:disable-next-line:quotemark
  return (fs.readFileSync(filePathInDisk, 'utf8') as string).replace(/"/g, "'");
}
