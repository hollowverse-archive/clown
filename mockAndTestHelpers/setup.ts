jest.mock('fs');
process.exit = jest.fn() as any;
jest.mock('../src/printErrors');
