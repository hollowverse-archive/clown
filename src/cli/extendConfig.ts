import path from 'path';
import { writeFile } from './utils';
import { computeFileContents } from './computeFileContents';
import _ from 'lodash';
import bluebird from 'bluebird';
import fs from 'fs-extra';

export async function extendConfig(cwd: string) {
  const fileContents = await computeFileContents(cwd);
  const iterableFileContents = _.entries(fileContents);

  await bluebird.each(
    iterableFileContents,
    async ([destinationPath, fileContent]) => {
      await fs.mkdirp(path.dirname(destinationPath));

      return writeFile(destinationPath, fileContent);
    },
  );
}
