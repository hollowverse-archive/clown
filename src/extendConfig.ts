import path from 'path';
import { writeFile } from './utils';
import { computeFileContents } from './computeFileContents';
import _ from 'lodash';
import bluebird from 'bluebird';
import fs from 'fs-extra';
import { isJsonEqual } from './isJsonEqual';

export async function extendConfig(cwd: string) {
  const fileContents = await computeFileContents(cwd);
  const iterableFileContents = _.entries(fileContents);

  await bluebird.each(
    iterableFileContents,
    async ([destinationPath, fileContent]) => {
      if (
        fileContent.destinationContent &&
        fileContent.type === 'json' &&
        isJsonEqual(fileContent.destinationContent, fileContent.computedContent)
      ) {
        return;
      }

      await fs.mkdirp(path.dirname(destinationPath));

      return writeFile(destinationPath, fileContent.computedContent);
    },
  );
}
