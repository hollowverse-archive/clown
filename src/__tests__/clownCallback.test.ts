import { extendConfig } from '../extendConfig';
import { readJsonNoDoubleQuotes } from '../../mockAndTestHelpers/readJsonNoDoubleQuotes';
import { vol } from '@forabi/memfs';
import fs from 'fs';

describe('clownCallback.js', () => {
  beforeEach(() => {
    vol.reset();
    jest.resetModules();
  });

  it('can edit a json', async () => {
    const files = {
      '/clown.json': `{
        "extensions": [
          "/clownExtensionA",
          "/clownOverride"
        ]
      }`,
      '/clownExtensionA/package.json': `{
        "foo": "a",
        "bar": "b"
      }`,
      '/package.json': `{
        "name": "Stuff"
      }`,
    };

    // @ts-ignore
    files['/clownOverride/clownCallback.js'] = '';

    vol.fromJSON(files);

    jest.doMock(
      '/clownOverride/clownCallback.js',
      () => (clownFs: any) => {
        return clownFs.editJson('/package.json', (pkgJson: any) => {
          delete pkgJson.foo;

          return pkgJson;
        }).fileContents;
      },
      { virtual: true },
    );

    await extendConfig('/');

    expect(readJsonNoDoubleQuotes('/package.json')).toMatchSnapshot();
  });

  // tslint:disable
  it('can delete a file', async () => {
    const files = {
      '/clown.json': `{
        "extensions": [
          "/clownExtensionA",
          "/clownOverride"
        ]
      }`,
      '/clownExtensionA/newJsonFile.json': `{
        "foo": "a",
        "bar": "b"
      }`,
    };

    // @ts-ignore
    files['/clownOverride/clownCallback.js'] = '';

    vol.fromJSON(files);

    jest.doMock(
      '/clownOverride/clownCallback.js',
      () => (clownFs: any) => {
        return clownFs.deleteFile('/newJsonFile.json').fileContents;
      },
      { virtual: true },
    );

    await extendConfig('/');

    expect(fs.existsSync('/newJsonFile.json')).toBe(false);
  });
});
