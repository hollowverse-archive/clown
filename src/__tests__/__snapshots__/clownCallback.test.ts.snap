// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clownCallback.js can edit a json 1`] = `
"{
  'name': 'Stuff',
  'bar': 'b'
}
"
`;
